### PR TITLE
Feature/hero filtering

### PR DESCRIPTION
**Pull request by k-szewczyk**
Added filtering and ordering, and a README file. You can also search for available opponents for a specified hero.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
...
+Rules are simple:
+ - Heroes can fight with each one only once,
+ - Only GrandMaster can choose attendees for every fight,
+ - Some races can't fight with another (for example hedgehogs can fight with hamsters, but can't fight with humans)
+
+
+###Prerequisites
+
+ - python 3.7
+ - docker, and docker-compose
```
Comment (natkam) [inline README.md]: Actually, you also need `Make` - you could also add it to the list.

File: backend/halloffame/filters.py [hunk 2]
```diff
@@ -1,12 +1,23 @@
...
+    find_opponents_for = filters.ModelChoiceFilter(method='available_opponents', label="find opponents for",
+                         queryset=Hero.objects.all())
+    is_alive = filters.BooleanFilter(label='is alive')
 
     class Meta:
         model = Hero
-        fields = ('level', 'race', 'guild', 'user_name')
+        fields = ('user_name', 'level', 'race', 'guild', 'is_alive', 'find_opponents_for')
+
+    def available_opponents(self, queryset, name, value: Hero):
```
Comment (natkam) [inline backend/halloffame/filters.py]: You could use a verb in the method name, e.g. `get_available_opponents`.

File: backend/halloffame/tests.py [hunk 3]
```diff
@@ -1,59 +1,106 @@
...
+        self.user1_url = reverse("hero-detail", args=(self.user1.id,))
+        self.user2 = UserFactory()
         self.user2_url = reverse("hero-detail", args=(self.user2.id,))
         self.client.force_login(self.user1)
 
     def test_create_new_hero(self):
-        response = self.client.post(self.hero_list_url,
-                                    {'user': self.user1.id, 'race': self.race1.id, 'guild': self.guild1.id})
+        self.client.post(self.hero_list_url,
+                         {'user': self.user1.id, 'race': self.race1.id, 'guild': self.guild1.id})
```
Comment (natkam) [inline backend/halloffame/tests.py]: It's not a bad idea to check the response status before checking other things:
```
from rest_framework import status
...
def test_create_new_hero(self):
    data = {'user': self.user1.id, 'race': self.race1.id, 'guild': self.guild1.id})
    response = self.client.post(self.hero_list_url, data=data)

    self.assertEqual(response.status_code, status.HTTP_201_CREATED)
    self.assertEqual(Hero.objects.count(), 1)
```

File: backend/halloffame/tests.py [hunk 4]
```diff
@@ -1,59 +1,106 @@
...
+        cls.battle2.looser = cls.hero1
+        cls.battle2.save()
+
+        cls.battle3 = Battle.objects.create(date=timezone.now(), is_looser_dead=True)
+        cls.battle3.attendees.set([cls.hero2, cls.hero3])
+        cls.battle3.looser = cls.hero3
+        cls.battle3.save()
+
+    def test_alive_heroes_filtering(self):
+        response = self.client.get(self.hero_list_url + '?is_alive=True')
```
Comment (natkam) [inline backend/halloffame/tests.py]: You could also do it this way:
```suggestion
        alive_heroes_url = reverse('hero-list', kwargs={'is_alive': 'True'}
        response = self.client.get(alive_heroes_url)
```

File: backend/halloffame/tests.py [hunk 5]
```diff
@@ -1,59 +1,106 @@
...
+        self.assertEqual(len(response.data['results']), 2)
+
+    def test_dead_heroes_filtering(self):
+        response = self.client.get(self.hero_list_url + '?is_alive=False')
+        self.assertEqual(len(response.data['results']), 1)
+
+    def test_available_opponents_for_specified_hero_filtering(self):
+        self.hero1.race.can_fight_with.set([self.hero1.race, self.hero2.race, self.hero3.race])
+        self.hero1.race.save()
+        hero4 = Hero.objects.create(user=UserFactory(), race=self.hero1.race)
```
Comment (natkam) [inline backend/halloffame/tests.py]: Use `HeroFactory`:
```suggestion
        hero4 = HeroFactory(race=self.hero1.race)
```

File: backend/halloffame/tests.py [hunk 6]
```diff
@@ -1,59 +1,106 @@
...
+
+class HallOfFameFilteringTestCase(test.APITestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.hero_list_url = reverse('hero-list')
+        cls.hero1 = HeroFactory()
+        cls.hero2 = HeroFactory()
+        cls.hero3 = HeroFactory()
+
+        cls.battle1 = Battle.objects.create(date=timezone.now())
```
Comment (natkam) [inline backend/halloffame/tests.py]: Don't you have a battle factory?

File: backend/halloffame/views.py [hunk 7]
```diff
@@ -8,9 +10,11 @@
 
 class HeroViewSet(ModelViewSet):
     permission_classes = (IsOwnerOrReadOnly,)
+    filter_backends = (filters.OrderingFilter, DjangoFilterBackend)
```
Comment (natkam) [inline backend/halloffame/views.py]: You could put all the filter-related attributes next to each other, it'd be easier to read.

File: backend/Sakaar/settings.py [hunk 8]
```diff
@@ -29,17 +29,18 @@
 # Application definition
 
 INSTALLED_APPS = [
+    'battles',
```
Comment (Mateusz-Slisz) [inline backend/Sakaar/settings.py]: As I can see you tried to alphabetical order installed apps, in my opinion, it should be arranged like this:
1. Django
2. external packages
3. apps

File: backend/halloffame/filters.py [hunk 9]
```diff
@@ -1,12 +1,23 @@
...
+                         queryset=Hero.objects.all())
+    is_alive = filters.BooleanFilter(label='is alive')
 
     class Meta:
         model = Hero
-        fields = ('level', 'race', 'guild', 'user_name')
+        fields = ('user_name', 'race', 'guild', 'is_alive', 'find_opponents_for')
+
+    def get_available_opponents(self, queryset, name, value: Hero):
+        hero_1 = value
```
Comment (Mateusz-Slisz) [inline backend/halloffame/filters.py]: unnecessary variable

File: backend/battles/models.py [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
-
+from django.utils import timezone
 
 class Battle(models.Model):
     attendees = models.ManyToManyField('halloffame.Hero', related_name='battles')
     looser = models.ForeignKey('halloffame.Hero', blank=True, null=True, default=None, related_name='lost_battles',
                                on_delete=models.SET_NULL)
     is_looser_dead = models.BooleanField(default=False)
-    date = models.DateTimeField()
+    date = models.DateTimeField(default=timezone.now())
```
Comment (Rafal-Maszlej) [inline backend/battles/models.py]: You shouldn't call a function here
https://stackoverflow.com/a/2771701

File: backend/halloffame/urls.py [hunk 11]
```diff
@@ -2,12 +2,12 @@
...
 from rest_framework_swagger.views import get_swagger_view
 
-from . import views
+from halloffame.views import HeroViewSet
 
 schema_view = get_swagger_view(title='Sakaar')
 
 router = routers.DefaultRouter()
-router.register('heroes', views.HeroViewSet)
+router.register('heroes', HeroViewSet)
```
Comment (Rafal-Maszlej) [inline backend/halloffame/urls.py]: Some `basename`? ;-)

Comment (natkam) [reply backend/halloffame/urls.py]: The basename is set to the queryset model name, lower-case, by default :) So it's not necessary here.

File: backend/halloffame/filters.py [hunk 12]
```diff
@@ -1,12 +1,22 @@
...
     class Meta:
         model = Hero
-        fields = ('level', 'race', 'guild', 'user_name')
+        fields = ('user_name', 'race', 'guild', 'is_alive', 'find_opponents_for')
+
+    def get_available_opponents(self, queryset, name, value: Hero):
+        available_heroes = Hero.objects.get_annotations().filter(
+            is_alive=True, race__in=value.race.can_fight_with.all()).exclude(user=value.user).exclude(
+            battles__in=value.battles.all()
+        )
```
Comment (natkam) [inline backend/halloffame/filters.py]: I'd say that the way the parentheses were placed and indented here, is somewhat confusing. The closing parenthesis looks like it matches the one after `filter` in line 18, and it's surprising when you see that there's `exclude` *inside* `filter`. Generally it takes a while to figure out which parenthesis belongs to which method. I'd suggest splitting the line using `\` or Python's implied line continuation inside parentheses, e.g. like this:
```
available_heroes = Hero.objects.get_annotations().filter(
    is_alive=True, race__in=value.race.can_fight_with.all()
).exclude(user=value.user).exclude(battles__in=value.battles.all())
```
Anyway, if you want to put the closing parenthesis in a new line, usually we follow the rule that its indentation should be the same as the line's in which the parenthesis was open, and whatever's inside the parentheses, should be additionally indented.

File: backend/halloffame/tests.py (hunk 6 above)
Comment (natkam) [reply backend/halloffame/tests.py]: So? :grin:

File: backend/halloffame/tests.py (hunk 4 above)
Comment (natkam) [reply backend/halloffame/tests.py]: ?

File: backend/halloffame/views.py [hunk 13]
```diff
@@ -1,16 +1,27 @@
...
 from halloffame.models import Hero
 from halloffame.permissions import IsOwnerOrReadOnly
 from halloffame.serializers import HeroSerializer
+from battles.models import Battle
 
 
 class HeroViewSet(ModelViewSet):
+    """This endpoint allows you to retrieve information about heroes
+
+     user_name -- returns hero filtered by username
```
Comment (natkam) [inline backend/halloffame/views.py]: Do you mean that one can filter users by username with such querystring? Like `api/heroes?user_name=Krzysiek`?

File: backend/halloffame/tests.py (hunk 6 above)
Comment (k-szewczyk) [reply backend/halloffame/tests.py]: I tried to write battle factory but I've run into problems with it. I will try to handle it today

File: backend/halloffame/views.py (hunk 13 above)
Comment (k-szewczyk) [reply backend/halloffame/views.py]: Whoops  :c I was testing how swagger documentation works, and I've forgot to delete it

File: backend/halloffame/filters.py (hunk 12 above)
Comment (k-szewczyk) [reply backend/halloffame/filters.py]: Okay, that looks better, just one question. Shouldn't be the last line indented to? Just like
```
available_heroes = Hero.objects.get_annotations().filter(
    is_alive=True, race__in=value.race.can_fight_with.all()
    ).exclude(user=value.user).exclude(battles__in=value.battles.all())
```

File: backend/halloffame/views.py (hunk 13 above)
Comment (k-szewczyk) [reply backend/halloffame/views.py]: But, anyway yea it works like that :)

File: backend/halloffame/filters.py (hunk 12 above)
Comment (natkam) [reply backend/halloffame/filters.py]: Well, there's a dozen styles and conventions and I myself find it rather ambiguous, and I'm not sure that one style is always better than others, so... It all depends ;) 
As to your question, [PEP 8 states](https://www.python.org/dev/peps/pep-0008/#indentation) that both styles are OK (the closing parenthesis may be indented or not), but - as I've mentioned - I have the impression that not indenting the last line is more common, at least here.
You could also experiment with breaking the line using `\`, like e.g. this:
```
available_heroes = Hero.objects.get_annotations()\
    .filter(is_alive=True, race__in=value.race.can_fight_with.all())\
    .exclude(user=value.user)\
    .exclude(battles__in=value.battles.all())
```
Actually, maybe it'd be even easier to understand this way :thinking:

File: backend/halloffame/views.py (hunk 13 above)
Comment (natkam) [reply backend/halloffame/views.py]: You could add a more verbose comment, it's a good idea to document endpoints like this! :smile:

File: backend/halloffame/tests.py (hunk 4 above)
Comment (natkam) [reply backend/halloffame/tests.py]: Ooops, actually I made a mistake, the extra arguments go in the `get` method and not `reverse`, here's the way to do it:
```
alive_heroes_url = reverse('hero-list')
response = self.client.get(alive_heroes_url, {'is_alive': 'True'})
```
Let me know if it works, and sorry for the confusion, my bad! :smile:

File: backend/halloffame/tests.py (hunk 6 above)
Comment (k-szewczyk) [reply backend/halloffame/tests.py]: Battle factory handled :muscle:

File: backend/halloffame/tests.py [hunk 14]
```diff
@@ -73,29 +74,19 @@ def setUpTestData(cls):
...
-        cls.battle3.looser = cls.hero3
-        cls.battle3.save()
+        cls.battle1 = BattleFactory(attendees=(cls.hero1, cls.hero2), looser=cls.hero1, is_looser_dead=False)
+        cls.battle2 = BattleFactory(attendees=(cls.hero1, cls.hero3), looser=cls.hero1, is_looser_dead=False)
+        cls.battle3 = BattleFactory(attendees=(cls.hero2, cls.hero3), looser=cls.hero3, is_looser_dead=True)
 
     def test_alive_heroes_filtering(self):
-        response = self.client.get(self.hero_list_url, kwargs={'is_alive': True})
-
+        alive_heroes_url = reverse('hero-list')
```
Comment (natkam) [inline backend/halloffame/tests.py]: This url could be moved to `setUpTestData`.

File: backend/halloffame/tests.py [hunk 15]
```diff
@@ -104,11 +95,9 @@ def test_available_opponents_for_specified_hero_filtering(self):
         self.hero1.race.can_fight_with.set([self.hero1.race, self.hero2.race, self.hero3.race])
```
Comment (natkam) [inline backend/halloffame/tests.py]: You have to be careful in the tests, no to edit anything you've created in `setUpTestData` - actually, here it doesn't matter, but in general if you want to edit objects, they should be created in `setUp`. Or, well, this assignment of races to fight with could be also done in the `setUpTestData` :)

**Pull request merged**